src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -3822,7 +3822,8 @@ void StorageReplicatedMergeTree::alter(
                 zookeeper_path + "/block_numbers", "block-", zookeeper_path + "/temp", *zookeeper);
 
             for (const auto & lock : lock_holder->getLocks())
-                mutation_entry.block_numbers[lock.partition_id] = lock.number;
+                if (mutation_entry.commands.isPartitionAffected(lock.partition_id, *this, query_context))
```

Comment:
It is easy to tell that current mutation command is partition-scoped by looking at its `partition` public field (no need to filter partitions otherwise). You should then be able to obtain partition_id by calling `MergeTreeData::getPartitionIDFromQuery()`. (As you currently do in `MutationCommands::isPartitionAffected()`)

Reply:
I don't understand here what that simplicity gives us. If any of commands doesn't have `partition` then we immediately add this block number. :man_shrugging: